### PR TITLE
v5.0.0 - Overall project restructuring & refactoring

### PR DESCRIPTION
**Pull request by cmyui**
# v5.0.0 - Overall project restructuring & refactoring
This one comes with quite a few updates and improvements to the codebase.
(implementing more of my learnings from time away)

## decoupling of logic & data into separate (functional) api layers
One of the main issues facing the codebase is it's tight coupling of logic & data.

Let's take the Beatmap class as an example - it;
- defines [the representation of an osu! beatmap](https://github.com/osuAkatsuki/bancho.py/blob/master/app/objects/beatmap.py#L272-L362)
- provides [methods to interact with specific instances](https://github.com/osuAkatsuki/bancho.py/blob/master/app/objects/beatmap.py#L375-L558)
- contains [the logical functionality for managing osu! beatmaps](https://github.com/osuAkatsuki/bancho.py/blob/master/app/objects/beatmap.py#L616-L768)

If you look through bancho.py's codebase, you'll see this recurring theme implemented in many inconsistent ways.
(see if you can find the similarities & differences in [the Player class](https://github.com/osuAkatsuki/bancho.py/blob/master/app/objects/player.py#L195-L1169) for example)

This PR aims to decouple structural representation, data providers, and business logic into 3 separate api levels.

### data structures will be represented as "models"
Models represent the format of a resource.
They allow you to think only about the structure of a resource - the attributes within it and its properties.

An example of a model may be the representation of a beamtap - for example;
```py
from dataclasses import dataclass

@dataclass
class Beatmap:
    id: int
    set_id: int
    md5: str
    ranked_status: int

    circle_size: float
    health_drain: float
    approach_rate: float
    overall_difficulty: float
```
(you may also define simple properties, but nothing involving complex logic or external data points)

### data providers will be represented as "repositories"
Repositories are an abstraction around the interaction with a resource.
They allow you to think of the interactions with the data providing service, as opposed to the actual implementation details of the underlying services - proving a standardized API for your business-logic layers.

Continuing with the theme of beatmaps, a simple repository may look something like
```py
from typing import Mapping
from typing import Optional

import models
import services

## in-memory cache

id_cache: Mapping[int, models.Beatmap] = {}
md5_cache: Mapping[str, models.Beatmap] = {}

## reads

async def fetch_by_id(id: int) -> Optional[models.Beatmap]:
    if beatmap := id_cache.get(id):
        return beatmap

    beatmap_row = await services.database.fetch_one(
        "SELECT * FROM beatmaps WHERE id = :id",
        {"id": id},
    )
    if beatmap_row is None:
        return None

    beatmap = models.Beatmap(**beatmap_row)

    id_cache[beatmap.id] = beatmap

    return beatmap

## writes

## updates

async def update_status(id: int, new_status: int) -> None:
    await services.database.execute(
        "UPDATE beatmaps SET status = :status WHERE id = :id",
        {"status": new_status, "id": id},
    )

    id_cache[id].status = new_status


async def update_playcount(id: int, playcount: int) -> None:
    await services.database.execute(
        "UPDATE beatmaps SET playcount = :playcount WHERE id = :id",
        {"playcount": playcount, "id": id},
    )

    id_cache[id].playcount = playcount

## deletes

async def delete_by_id(id: int) -> None:
    await services.database.execute(
        "DELETE FROM beatmaps WHERE id = :id",
        {"id": id},
    )

    if id in id_cache:
        del id_cache[id]
```
(you may also have functions to fetch by other means, update, or create beatmaps)

### business logic will be represented by specific "usecases"
Usecases define the logical interface exposed by a resource.
They allow you to interact with & make complex calculations & modifications possibly depending on multiple resources simultaneously.

Building on our previous examples further, some use cases of a beatmap may look something like
```py
import models
import repositories

async def update_status(beatmap: models.Beatmap, new_status: int) -> None:
    await repositories.beatmaps.update_status(beatmap.id, new_status)


async def update_playcount(beatmap: models.Beatmap) -> None:
    await repositories.beatmaps.update_playcount(beatmap.id, beatmap.plays)
```

Splitting the objects into layers like this allows us to implement a standardized API across the codebase's multiple resources.

With this sort of orthogonal design, we can increase code reuse significantly, increase the level of abstraction the programmer works at, and use this api as a framework for writing unit & smoke tests for the regular user flow in the application trivially.

## restful api restructuring
This PR also resolves the paths provided by the v1 API to be restful -- we've updated the paths to fit a resource-driven design.
```
# GET /players/{player_id}
# GET /players/{player_id}/stats
# GET /players/{player_id}/status
# GET /players/{player_id}/scores/best
# GET /players/{player_id}/scores/recent
# GET /players/{player_id}/most-played-beatmaps
# GET /players/leaderboard

# GET /beatmaps/{beatmap_id}
# GET /beatmaps/{beatmap_id}/scores

# GET /scores/{score_id}
# GET /scores/{score_id}/replay

# GET /matches/{match_id}
```

## assume developers will use cloudflare (flex) for ssl
This simplifies the server setup itself quite significantly - no need for an ssl certificate, complex nginx setup or anything - just listen on port 80 and have cloudflare do the rest - [implemented here](https://github.com/osuAkatsuki/bancho.py/pull/205/commits/c02ad12b9f4fba4c54aed6b97938341ddca8fefd).

## disclaimer
Note that the examples provided in this are simplifications, and the exact implementation details of the PR are still being actively worked on - nothing is final!

### CODE REVIEW

**Comment by cmyui** [comment]
the code is getting much nicer

File: .env.example [hunk 1]
```diff
@@ -10,7 +10,7 @@ REDIS_DSN=redis://localhost
 OSU_API_KEY=
 
 DOMAIN=cmyui.xyz
-MIRROR_URL=https://api.chimu.moe/v1
+MIRROR_URL=https://catboy.best/api
```
Comment (cmyui) [inline .env.example:13]: why does the current most reliable beatmap mirror need to be named catboy.best

Comment (def750) [reply .env.example:13]: is good

**Comment by alowave223** [comment]
Absolutely love how's cmyui is learning new things and me catchig them from him, good work and i really like this new api, really restfull and easy to understand, also the usecases makes feel code more consistant and logicaly sorted(?), i mean you can easy understand what file you need to edit for that module or function, very cool. Keep it up!

File: .env.example (hunk 1 above)
Comment (JKBGL) [reply .env.example:13]: > Can you add reminding that you can still use chimu ? 👉👈

You kinda took too long fixing it, so everyone started migrating to mino..

Comment (KotRikD) [reply .env.example:13]: > > Can you add reminding that you can still use chimu ? 👉👈
> 
> You kinda took too long fixing it, so everyone started migrating to mino..

tbh, its not easy task to move "our big dump" from ipfs to raw FS :/ That's one of the reason why it took sooo long. And now chimu should have 99.99% uptime :)

**Comment by cmyui** [comment]
you ever just work on a pr for over a month?

**Comment by cmyui** [comment]
i may split up account & session objects in this PR?

**Comment by 7ez** [comment]
> you ever just work on a pr for over a month?

2 months now :trollface:

**Comment by cmyui** [comment]
this is gonna be a hell of a merge conflict

**Comment by cmyui** [comment]
oh wait, i already merged above

thanks previous me!

**Comment by cmyui** [comment]
p.s. this is not how you're supposed to use pull requests 🥲

**Comment by cmyui** [comment]
hm the example i gave for beatmaps doesn't work so well in the usecase since the usecases i showed are just 1:1 with the repositories lol.. maybe will think of a better example

**Comment by cmyui** [comment]
while this has helped with orienting my vision for the project more accurately - it's not worth the time to try finishing this at the moment

i'll slowly be merging the ideas explored in this pr into the master branch in smaller, more atomized parts over time - the motivations still hold strong

**Pull request closed without merging**